src/mod_mam_elasticsearch_arch.erl

```diff
@@ -0,0 +1,291 @@
+                       [Host, Err]),
+            Err
+    end.
+
+-spec archive_size(Size :: integer(),
+                   Host :: jid:server(),
+                   _ArchiveId,
+                   OwnerJid :: jid:jid()) -> non_neg_integer().
+archive_size(_Size, _Host, _ArchiveId, OwnerJid) ->
+    SearchQuery = build_search_query(#{owner_jid => OwnerJid}),
+    archive_size(SearchQuery).
+
+-spec remove_archive(Acc,
+                     Host :: jid:server(),
+                     ArchiveId :: mod_mam:archive_id(),
+                     OwnerJid :: jid:jid()) -> Acc when Acc :: map().
+remove_archive(Acc, Host, _ArchiveId, OwnerJid) ->
+    SearchQuery = build_search_query(#{owner_jid => OwnerJid}),
+    case mongoose_elasticsearch:delete_by_query(?INDEX_NAME, ?TYPE_NAME, SearchQuery) of
+        ok ->
+            ok;
+        {error, _} = Err ->
+            ?ERROR_MSG("event=remove_archive_failed server=~s user=~s reason=~1000p",
+                       [Host, jid:to_binary(OwnerJid), Err]),
+            ok
+    end,
+    Acc.
+
+%%-------------------------------------------------------------------
+%% Helpers
+%%-------------------------------------------------------------------
+
+-spec hooks(jid:lserver()) -> [ejabberd_hooks:hook()].
+hooks(Host) ->
+    [{mam_archive_message, Host, ?MODULE, archive_message, 50},
+     {mam_lookup_messages, Host, ?MODULE, lookup_messages, 50},
+     {mam_archive_size, Host, ?MODULE, archive_size, 50},
+     {mam_remove_archive, Host, ?MODULE, remove_archive, 50}].
+
+-spec make_document_id(binary(), mod_mam:message_id()) -> binary().
+make_document_id(Owner, MessageId) ->
+    <<Owner/binary, $$, (integer_to_binary(MessageId))/binary>>.
+
+-spec make_document(mod_mam:message_id(), binary(), binary(), binary(), exml:element()) ->
+    map().
+make_document(MessageId, Owner, Remote, SourceBinJid, Packet) ->
+    #{mam_id     => MessageId,
+      owner      => Owner,
+      remote     => Remote,
+      source_jid => SourceBinJid,
+      message    => exml:to_binary(Packet),
+      body       => exml_query:path(Packet, [{element, <<"body">>}, cdata])
+     }.
+
+-spec build_search_query(map()) -> mongoose_elasticsearch:query().
+build_search_query(Params) ->
+    Filters = build_filters(Params),
+    TextSearchQuery = build_text_search_query(Params),
+    #{query =>
+      #{bool =>
+        #{must => TextSearchQuery,
+          filter => Filters}}}.
+
+-spec build_filters(map()) -> [map()].
+build_filters(Params) ->
+    Builders = [fun owner_filter/1,
+                fun with_jid_filter/1,
+                fun range_filter/1],
+    lists:flatmap(fun(F) -> F(Params) end, Builders).
+
+-spec owner_filter(map()) -> [map()].
+owner_filter(#{owner_jid := Owner}) ->
+    BinOwner = mod_mam_utils:bare_jid(Owner),
+    [#{term => #{owner => BinOwner}}].
+
+-spec with_jid_filter(map()) -> [map()].
+with_jid_filter(#{with_jid := #jid{} = WithJid}) ->
+    [#{term => #{remote => mod_mam_utils:bare_jid(WithJid)}}];
```

Comment:
this line is not covered by tests

Reply:
Test which covers it was excluded from mam_SUITE. I've added it again, let's see if it'll pass for other backends.